src/hotspot/share/memory/archiveBuilder.cpp

```diff
@@ -593,6 +595,226 @@ void ArchiveBuilder::make_klasses_shareable() {
+
+      switch (type) {
+      case MetaspaceObj::ClassType:
+        write_klass((Klass*)src, runtime_dest, type_name, bytes, THREAD);
+        break;
+      case MetaspaceObj::ConstantPoolType:
+        write_klass(((ConstantPool*)src)->pool_holder(),
+                    runtime_dest, type_name, bytes, THREAD);
+        break;
+      case MetaspaceObj::ConstantPoolCacheType:
+        write_klass(((ConstantPoolCache*)src)->constant_pool()->pool_holder(),
+                    runtime_dest, type_name, bytes, THREAD);
+        break;
+      case MetaspaceObj::MethodType:
+        write_method((Method*)src, runtime_dest, type_name, bytes, THREAD);
+        break;
+      case MetaspaceObj::ConstMethodType:
+        write_method(((ConstMethod*)src)->method(), runtime_dest, type_name, bytes, THREAD);
+        break;
+      case MetaspaceObj::SymbolType:
+        {
+          ResourceMark rm(THREAD);
+          Symbol* s = (Symbol*)src;
+          log_debug(cds, map)(_LOG_PREFIX " %s", p2i(runtime_dest), type_name, bytes,
+                              s->as_quoted_ascii());
+        }
+        break;
+      default:
+        log_debug(cds, map)(_LOG_PREFIX, p2i(runtime_dest), type_name, bytes);
+        break;
+      }
+
+      last_obj_base = dest;
+      last_obj_end  = dest + bytes;
+    }
+
+    write_data(last_obj_base, last_obj_end, last_obj_base + buffer_to_runtime_delta());
+    if (last_obj_end < region_end) {
+      log_debug(cds, map)(PTR_FORMAT ": @@ Misc data " SIZE_FORMAT " bytes",
+                          p2i(last_obj_end + buffer_to_runtime_delta()),
+                          size_t(region_end - last_obj_end));
+      write_data(last_obj_end, region_end, last_obj_end + buffer_to_runtime_delta());
+    }
+  }
+
+#undef _LOG_PREFIX
+
+  // Write information about a region, whose address at dump time is [base .. top). At
+  // runtime, this region will be mapped to runtime_base.  runtime_base is 0 if this
+  // region will be mapped at os-selected addresses (such as the bitmap region), or will
+  // be accessed with os::read (the header).
+  void write_region(const char* name, address base, address top, address runtime_base) {
+    size_t size = top - base;
+    base = runtime_base;
+    top = runtime_base + size;
+    log_info(cds, map)("[%-18s " PTR_FORMAT " - " PTR_FORMAT " " SIZE_FORMAT_W(9) " bytes]",
+                       name, p2i(base), p2i(top), size);
+  }
+
+  // open and closed archive regions
+  void write_heap_region(const char* which, GrowableArray<MemRegion> *regions) {
+    for (int i = 0; i < regions->length(); i++) {
+      address start = address(regions->at(i).start());
+      address end = address(regions->at(i).end());
+      write_region(which, start, end, start);
+      write_data(start, end, start);
+    }
+  }
+
+  // Dump all the data [base...top). Pretend that the base address
+  // will be mapped to runtime_base at run-time.
+  static void write_data(address base, address top, address runtime_base) {
```

Comment:
Consider printing the data with os::print_hex_dump() instead. Would slim down this coding.